cvmfs/receiver/catalog_merge_tool.h

```diff
@@ -0,0 +1,94 @@
+/**
+ * This file is part of the CernVM File System.
+ */
+
+#ifndef CVMFS_RECEIVER_CATALOG_MERGE_TOOL_H_
+#define CVMFS_RECEIVER_CATALOG_MERGE_TOOL_H_
+
+#include <string>
+#include <vector>
+
+#include "catalog_diff_tool.h"
+#include "catalog_mgr_rw.h"
+#include "params.h"
+#include "util/pointer.h"
+
+namespace catalog {
+class WritableCatalogManager;
+}
+
+namespace download {
+class DownloadManager;
+}
+
+namespace manifest {
+class Manifest;
+}
+
+namespace shash {
+class Any;
+}
+
+namespace receiver {
+
+class CatalogMergeTool : public CatalogDiffTool {
+ public:
+  CatalogMergeTool(const std::string& repo_path,
+                   const std::string& old_root_hash,
+                   const std::string& new_root_hash,
+                   const std::string& temp_dir_prefix,
+                   download::DownloadManager* download_manager,
+                   manifest::Manifest* manifest);
+  virtual ~CatalogMergeTool();
+
+  bool Run(const Params& params, std::string* new_manifest_path);
+
+ protected:
+  virtual void ReportAddition(const PathString& path,
+                              const catalog::DirectoryEntry& entry,
+                              const XattrList& xattrs);
+  virtual void ReportRemoval(const PathString& path,
+                             const catalog::DirectoryEntry& entry);
+  virtual void ReportModification(const PathString& path,
+                                  const catalog::DirectoryEntry& old_entry,
+                                  const catalog::DirectoryEntry& new_entry);
+
+ private:
+  struct ChangeItem {
+    enum ChangeType { kAddition, kRemoval, kModification };
+    ChangeItem(ChangeType type, const PathString& path,
+               const catalog::DirectoryEntry& entry1);
+    ChangeItem(ChangeType type, const PathString& path,
+               const catalog::DirectoryEntry& entry1, const XattrList& xattrs);
+    ChangeItem(ChangeType type, const PathString& path,
+               const catalog::DirectoryEntry& entry1,
+               const catalog::DirectoryEntry& entry2);
+    ChangeItem(const ChangeItem& other);
```

Comment:
I don't understand this. I actually need the copy constructor and assignment operator, since I'm putting objects of `ChangeItem` in a `std::vector`. I added an explicit definition of `operator=`.